Add new feature to find file

koreamic opened the pull request:
Using the regular expression, develop "file finder" feature.
(http://feedback.gitlab.com/forums/176466-general/suggestions/4987909-add-file-finder-fuzzy-input-in-files-tab-to-ju)

Issue(#9854)

stanhu commented:
Thank you for submitting this PR! There are a number of rubocop errors that you can probably fix quickly. Can you post screenshots of how this feature will look?

koreamic committed:
Add new feature to find file

Using the regular expression, develop "file finder" feature.
http://feedback.gitlab.com/forums/176466-general/suggestions/4987909-add-file-finder-fuzzy-input-in-files-tab-to-ju

koreamic commented:
Thank you for your concern!!

Please see gif ficture bellow for reviewing this feature.

Thank you,

(I fixed rubocop errors and I did commit & push in my fork repository )

![gitlab_file_finder_2015_11_26_20_35_26_240](https://cloud.githubusercontent.com/assets/14069553/11422245/c3d8db7a-947d-11e5-8fe2-236dd3e83ea5.gif)

stanhu commented on app/controllers/projects/tree_filter_controller.rb:
```diff
@@ -0,0 +1,20 @@
+  before_action :assign_ref_vars
+
+  def show
+    return render_404 unless @repository.commit(@ref)
```
Do you need to check whether the current user has permission to this repository?

DouweM replied:
I think we do, like in the TreeController.

stanhu commented on app/controllers/projects/tree_filter_controller.rb:
```diff
@@ -0,0 +1,20 @@
+  def show
+    return render_404 unless @repository.commit(@ref)
+
+    r = Grit::Repo.new(@repo.path_to_repo, { is_bare: true })
```
@DouweM Is GitLab still using Grit or should we be using Rugged here?

DouweM replied:
@stanhu Rugged all the way.

stanhu commented:
@koreamic Great work! This is a fantastic feature.

stanhu commented on app/assets/javascripts/project_tree_filter.js.coffee:
```diff
@@ -0,0 +1,54 @@
+class @ProjectTreeFilter
+  constructor: ->
+
+    #focus text input box
```
Please add a space between the `#` and the comment.

stanhu commented on app/assets/stylesheets/pages/tree.scss:
```diff
@@ -1,5 +1,14 @@
 .tree-holder {
 
+
+  .tree-filter{
```
Add a space before the `{`.

stanhu commented on app/controllers/projects/tree_filter_controller.rb:
```diff
@@ -0,0 +1,20 @@
+
+    respond_to do |format|
+      format.html
+      # Disable cache so browser history works
```
What is the issue here?

stanhu commented on app/views/projects/tree/show.html.haml:
```diff
@@ -3,12 +3,22 @@
+    %span.btn-group{class: 'btn-group pull-right hidden-xs hidden-sm', style: 'padding-right:2px;'}
+      = link_to namespace_project_tree_filter_path(@project.namespace, @project, @ref), class: 'btn btn-success', rel: 'nofollow' do
+        %i.fa.fa-filter
+        %span TREE FILTER
```
Should this be called "Find Files" instead?

Razer6 replied:
Should this be a dedicated? I would use `Mousetrap` to bind it to a keyboard shortcut instead. THis makes the UI less clumpsy and avoids using another click.

stanhu replied:
Good idea. Perhaps a keyboard shortcut when a user enters `f` may be the way to go.

Razer6 replied:
`t` is a good idea. With `ESC` back to the normal tree view.

DouweM replied:
@stanhu "Find File", singular, rather ;)

DouweM replied:
I want both a button and a `t` keyboard shortcut, the shortcut alone is not findable enough.

stanhu commented on app/assets/javascripts/project_tree_filter.js.coffee:
```diff
@@ -0,0 +1,54 @@
+        $(this).hide()
+
+  #mark text(awefwbwgtc -> <b>a</b>wefw<b>b</b>wgt<b>c</b> )
+  markText = (txt, charArr) ->
```
This is probably better named `highlightText`.

stanhu commented on app/assets/javascripts/project_tree_filter.js.coffee:
```diff
@@ -0,0 +1,54 @@
+    for char in charArr
+      charIdx = target.toLowerCase().indexOf char.toLowerCase()
+      result += "#{target.substring 0, charIdx}#{(target.charAt charIdx).bold()}"
+      target = target.substring charIdx+1
```
Add spaces: ```target = target.substring charIdx + 1```

stanhu commented on app/assets/javascripts/project_tree_filter.js.coffee:
```diff
@@ -0,0 +1,54 @@
+    return result
+
+  #generate regular expression(abc -> a[^b]*b[^c]*c)
+  makeRegEx = (searchCharArr) ->
```
What happens if the user enters `\` or some other special regex character?

DouweM replied:
Ping!

koreamic committed:
review

DouweM commented on config/routes.rb:
```diff
@@ -456,6 +456,15 @@
+            '/tree_filter/*id',
+            to: 'tree_filter#show',
+            constraints: { id: /.+/, format: /(html|js)/ },
+            as: :tree_filter
```
I prefer the name `find_file` to `tree_filter`.

DouweM commented on app/assets/javascripts/dispatcher.js.coffee:
```diff
@@ -103,6 +103,8 @@ class Dispatcher
         new Activities()
 
+    console.log("path: " + path)
+    console.log("path.first(): " + path.first())
```
Stray logging!

DouweM commented on app/views/projects/tree/show.html.haml:
```diff
@@ -3,12 +3,22 @@
 - if can? current_user, :download_code, @project
   .tree-download-holder
     = render 'projects/repositories/download_archive', ref: @ref, btn_class: 'btn-group pull-right hidden-xs hidden-sm', split_button: true
+  .tree-download-holder
```
Please use a different class here.

DouweM commented on app/views/projects/tree/show.html.haml:
```diff
@@ -3,12 +3,22 @@
   .tree-download-holder
     = render 'projects/repositories/download_archive', ref: @ref, btn_class: 'btn-group pull-right hidden-xs hidden-sm', split_button: true
+  .tree-download-holder
+    %span.btn-group{class: 'btn-group pull-right hidden-xs hidden-sm', style: 'padding-right:2px;'}
```
if this needs `padding-right`, please move that to the stylesheet.

DouweM commented on app/views/projects/tree/show.html.haml:
```diff
@@ -3,12 +3,22 @@
+  .tree-download-holder
+    %span.btn-group{class: 'btn-group pull-right hidden-xs hidden-sm', style: 'padding-right:2px;'}
+      = link_to namespace_project_tree_filter_path(@project.namespace, @project, @ref), class: 'btn btn-success', rel: 'nofollow' do
+        %i.fa.fa-filter
```
Please use the `icon` helper: `= icon('filter')`

DouweM replied:
Perhaps a search icon makes more sense than `filter`, by the way.

DouweM commented on app/views/projects/tree/show.html.haml:
```diff
@@ -3,12 +3,22 @@
+    %span.btn-group{class: 'btn-group pull-right hidden-xs hidden-sm', style: 'padding-right:2px;'}
+      = link_to namespace_project_tree_filter_path(@project.namespace, @project, @ref), class: 'btn', rel: 'nofollow' do
+        %i.fa.fa-filter
+        %span TREE FILTER
```
Do we need this button in the source twice?

DouweM commented on app/views/projects/tree_filter/_blob_item.html.haml:
```diff
@@ -0,0 +1,7 @@
+- tree_limit = 20 unless local_assigns[:tree_limit]
+
+%tr{ class: "tree-item #{blob_item_counter < tree_limit ? '' : 'hide'}" }
```
We shouldn't render every single file, since this can run into the thousands upon thousands for large repos (the Linux tree has 50k+)

DouweM commented on app/views/projects/tree_filter/show.html.haml:
```diff
@@ -0,0 +1,24 @@
+- tree_filter_list = @tree_filter_list
+- path = @path
+- id = @id
+- commit = @commit
```
Why do we need these local assignments?

DouweM commented on app/views/projects/tree_filter/show.html.haml:
```diff
@@ -0,0 +1,24 @@
+- id = @id
+- commit = @commit
+
+- page_title path.presence || "Files", ref
```
Since we don't actually take `@path` into account and always search from the root, I think we can just do `page_title "Find File", ref`

DouweM commented on app/controllers/projects/tree_filter_controller.rb:
```diff
@@ -0,0 +1,20 @@
+    return render_404 unless @repository.commit(@ref)
+
+    r = Grit::Repo.new(@repo.path_to_repo, { is_bare: true })
+    @tree_filter_list = r.lstree(@ref, recursive: true)
```
Can we do this using Rugged? And perhaps over AJAX so we don't slow down the initial request so much, and the JSON body can be cached?

DouweM commented on app/controllers/projects/tree_filter_controller.rb:
```diff
@@ -0,0 +1,20 @@
+    respond_to do |format|
+      format.html
+      # Disable cache so browser history works
+      format.js { no_cache_headers }
```
What do we use the JS format for?

DouweM commented on app/assets/javascripts/project_tree_filter.js.coffee:
```diff
@@ -0,0 +1,54 @@
+    regEx = makeRegEx searchCharArr
+
+    # file list count
+    showCnt = 20;
```
No semicolons please!

DouweM commented on app/assets/javascripts/project_tree_filter.js.coffee:
```diff
@@ -0,0 +1,54 @@
+  # filter tree
+  treeFilter: (e) ->
+    searchCharArr = $(this).val().split("")
+    regEx = makeRegEx searchCharArr
```
Would it be possible to use a more powerful fuzzy filtering algorithm like https://www.npmjs.com/package/fuzzaldrin-plus ?

DouweM commented on app/views/projects/tree_filter/_tree_filter_content.html.haml:
```diff
@@ -0,0 +1,6 @@
+%div.tree-content-holder
```
In the GIF, I notice there is too much padding on the left and right sides.

DouweM commented:
Would it be possible to add keyboard navigation of the results, like on GitHub?

DouweM commented on app/views/projects/tree/show.html.haml:
```diff
@@ -3,12 +3,22 @@
-
 = render 'projects/last_push'
 
 - if can? current_user, :download_code, @project
```
I don't think we need this check here at all, since the TreeController already verifies this.

koreamic commented:
Thank you for your(@DouweM, @stanhu, @Razer6) review.
Today's commit is not complete sources.
I seems to need some time for applying for your suggestions.
When I complete, reopen this PR!!

JobV commented:
@koreamic This is something we're very interested in, let us know if you need any help with completing this. We're happy to help.

koreamic commented:
@Job Wow!! Sincerely thank Thank you for your concern
I was cheered by your words.
I have been develop this feature. Actually I had some problems

1. Using Ajax in rails
2. Using Fuzzy Lib in browser
3. how to get a result like git command(git ls-tree --name-only or git ls-files) in Rugged
4. etc...

but now

Searching and studying, Ajax and Fuzzy Lib problem was solved (use browerify fuzzaldrin-plus).
Remained problem is that using Rugged, It is to get data like 'git ls-files'.

If you review my sources or suggest other good way, It is very helpful for me.
I will reopen this PR tomorrow.

Thank you again. have a nice day. ^^

DouweM commented:
@koreamic If getting the list of files using Rugged is not (easily) possible, another option is to adapt `Gitlab::Git::Repository`, which is part of our `gitlab_git` gem, and add a method that calls the `git ls-tree` command directly, parses the output and returns it in some manner that is useful to GitLab. This is also done for a number of other commands, like `log`: https://gitlab.com/gitlab-org/gitlab_git/blob/master/lib/gitlab_git/repository.rb#L215

You can then add a method on `Repository` within GitLab that calls this `Gitlab::Git::Repository` method, and use that method in your controller.
